File: solutions/JavaScript/valid_sudoku_hash_sets.js
function isValidSudoku(board) {
    const rows  = Array.from({ length: 9 }, () => new Set());
    const cols  = Array.from({ length: 9 }, () => new Set());
    const boxes = Array.from({ length: 9 }, () => new Set());

    for (let r = 0; r < 9; r++) {
        for (let c = 0; c < 9; c++) {
            const d = board[r][c];
            if (d === '.') continue;
            const box = Math.floor(r / 3) * 3 + Math.floor(c / 3);
            if (rows[r].has(d) || cols[c].has(d) || boxes[box].has(d)) return false;
            rows[r].add(d);
            cols[c].add(d);
            boxes[box].add(d);
        }
    }
    return true;
}

File: solutions/Python/valid_sudoku_hash_sets.py
from typing import List
from collections import defaultdict


def isValidSudoku(board: List[List[str]]) -> bool:
    rows = defaultdict(set)
    cols = defaultdict(set)
    boxes = defaultdict(set)

    for r in range(9):
        for c in range(9):
            d = board[r][c]
            if d == '.':
                continue
            box = (r // 3) * 3 + (c // 3)
            if d in rows[r] or d in cols[c] or d in boxes[box]:
                return False
            rows[r].add(d)
            cols[c].add(d)
            boxes[box].add(d)

    return True
